### PR TITLE
Resolves #355: Support grabbing the bytes of intervals

### LINKED ISSUE #355: Feature: support grabbing the bytes of intervals.
In some use cases one wants to parse a structure and also grab the raw input representation. This is sometimes wanted when one needs to hash a structure or check a signature: it's important to use the original bytes and not a re-serialised version which may canonicalise the representation.

It is already possible to get the current byte offset
```
peekByteOffset :: Decoder s ByteOffset
```
and this is used in [decodeWithByteSpan](https://hackage-content.haskell.org/package/cborg-0.2.10.0/docs/Codec-CBOR-Decoding.html#v:decodeWithByteSpan) to capture the pattern:
```
!before <- peekByteOffset
x <- decode
!after  <- peekByteOffset
```
And then it is possible as a post-processing pass to use these byte offsets to grab the spans of the original input. However doing this in two passes is slow and cumbersome.

What we would really like is to grab the input directly, something like this:
```
markInputSpan
x <- decode
x_bytes  <- getInputSpan
```
The rule here is that these things must be properly nested: the `markInputSpan` and `getInputSpan` must match up in pairs and be properly nested. This rule is just a suggestion because it probably enables a fast implementation more easily. If another approach does not require nesting, then that's ok.

The challenge however is to implement this so that it does not have a performance impact on users who do _not_ use the feature. In particular we cannot retain all input data buffers on the off chance that someone might want to grab some input.

Recall that the parser is incremental: it operates on one input buffer at a time. It asks for a new buffer when it runs out, and it forgets the previous input buffer (and most decoder primitives do not retain references to those input buffers to allow them to be GC'd).

One implementation idea is for the `go_slow` interpreter to maintain an `STRef` with the retained chunks, and a stack of input span marks. The `go_fast` would return to the slow path to handle this case, so it doesn't perturb the performance of the fast path. When the stack of input span marks is empty then the ` go_slow` interpreter would discard input buffers as it does now. But when it is non-empty it would retain them in some appropriate structure (probably along with the initial byte offset of each buffer) when each buffer would otherwise be exhausted and discarded. The implementation for `markInputSpan` would be to push a byte offset to the stack, and `getInputSpan` would pop and then use the retained buffers to extract the span.

Then the cost when the feature is not used, would be just the check when each input buffer is about to be discarded to see if it should instead be retained.

### PR DESCRIPTION
**Pull request by phadej**

### CODE REVIEW

**Comment by phadej** [comment]
I couldn't run macro benchmarks (#362), there are no differences in micro benchmarks

**Review by dcoutts — COMMENTED** [review]
Thoughts so far.

Generally this looks very promising: providing the feature at low cost to non-users.

File: cborg/src/Codec/CBOR/Read.hs [hunk 1]
```diff
@@ -1419,25 +1506,41 @@ getTokenVarLen len bs offset =
...
 
         | otherwise -> getTokenVarLenSlow
                          [bs',bs]
                          (len - (BS.length bs + BS.length bs'))
-                         offset
-
-getTokenVarLenSlow :: [ByteString] -> Int -> ByteOffset
-                   -> IncrementalDecoder s (ByteString, ByteString)
-getTokenVarLenSlow bss n offset = do
+                         (offset + intToInt64 (BS.length bs'))
```
[the remark this answers is not in the document]
Comment (phadej) [reply cborg/src/Codec/CBOR/Read.hs:1514]: > I've not followed why we're changing the offset calculations here

Because `getTokenVarLen/Slow` didn't use `offset` for anything, only to report an error. Previously (as you can see from the diff) it wasn't updated at all even after new chunks were read.

A side-effect of this change is that "end-of-input" error will no report the actualy end-of-input offset, previously it reported an offset at which `getTokenVarLen` was called.

File: cborg/src/Codec/CBOR/Read.hs [hunk 2]
```diff
@@ -1250,15 +1318,15 @@ go_slow da bs !offset = do
       mbs <- needChunk
       case mbs of
         Nothing   -> decodeFail bs' offset' "end of input"
-        Just bs'' -> go_slow da' bs'' offset'
+        Just bs'' -> go_slow da' bs'' offset' (slowMarkChunk bs'' offset' marks)
```
[the remark this answers is not in the document]
Comment (phadej) [reply cborg/src/Codec/CBOR/Read.hs:1321]: Well, I doubt there's any noticeable cost cause

```haskell
slowMarkChunk :: ByteString -> ByteOffset -> Marks -> Marks
slowMarkChunk _  _    NoMarks            = NoMarks
```

My gut feeling is there's more accumulated cost of threading an additional `NoMarks` argument through the interpreter.

File: cborg/tests/Tests/ByteOffset.hs [hunk 3]
```diff
@@ -63,6 +66,24 @@ prop_ATerm_isomorphic3 t =
...
+empty_deserialise :: Property
+empty_deserialise = res === (0, 0)
+  where
+    -- unfortunately we need a non-empty input stream.
+    -- the decoder will fail unconditionally, if there are commands,
+    -- even ones which consume nothing.
+    res = either throw snd (deserialiseFromBytes emptyDecoder (LBS.pack [0]))
+
+empty_deserialise_fail :: Bool
+empty_deserialise_fail = isLeft (deserialiseFromBytes emptyDecoder LBS.empty)
```
Comment (phadej) [inline cborg/tests/Tests/ByteOffset.hs:79]: I added these tests for both `peekByteOffset` and `getInputSpan`. 

I find it a bit surprising as `peekByteOffset` as well as `unmarkInput` and `getInputSpan` do succeed at the end of input, but they seem to fail at the very beginning of an empty input.

This is quite obscure corner case, but nevertheless I think it's good to be aware off (i.e. have test for it).

File: cborg/src/Codec/CBOR/Read.hs [hunk 4]
```diff
@@ -1198,48 +1207,107 @@ go_fast_end !bs (ConsumeMapLenOrIndef k) =
...
+      (bstr, bs'', marks') <- getTokenVarLen len bs' (offset' + intToInt64 (BS.length bs')) marks
       let !str = BA.fromByteString bstr
-      lift (k str) >>= \daz -> go_slow daz bs'' (offset' + intToInt64 len)
+      lift (k str) >>= \daz -> go_slow daz bs'' (offset' + intToInt64 len) marks'
       where
         !offset' = offset + intToInt64 (BS.length bs - BS.length bs')
 
     SlowConsumeTokenString bs' k len -> do
-      (bstr, bs'') <- getTokenVarLen len bs' offset'
+      (bstr, bs'', marks') <- getTokenVarLen len bs' (offset' + intToInt64 (BS.length bs')) marks
```
[the remark this answers is not in the document]
Comment (dcoutts) [reply cborg/src/Codec/CBOR/Read.hs:1299]: Got it. Thanks. And thanks for pushing the slight simplification.

File: cborg/src/Codec/CBOR/Read.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dcoutts) [reply cborg/src/Codec/CBOR/Read.hs:1321]: Indeed. And as you noted, the micro-benchmarks show nothing measureable. And I fixed the macro-benchmarks and they also show nothing measurable.

**Review by dcoutts — APPROVED** [review]
This is great. It adds the feature nicely, and doesn't add any cost for users not using it. And indeed using the feature itself is pretty cheap (as measured by a macro-benchmark).

I've taken the liberty of doing some renaming. And I'll add more API docs. I'll merge once I've finished adding docs.

File: cborg/src/Codec/CBOR/Decoding.hs [hunk 5]
```diff
@@ -1047,7 +1125,7 @@ peekByteSpan = Decoder (\k -> return (PeekByteSpan k))
 --
 -- > openByteSpan
 -- > x <- decode
--- > !after  <- peekByteSpan
+-- > bytes <- peekByteSpan
```
Comment (phadej) [inline cborg/src/Codec/CBOR/Decoding.hs:1128]: you removed bang here, but not in the implementation.

Comment (dcoutts) [reply cborg/src/Codec/CBOR/Decoding.hs:1128]: Thanks. I think the bang there is unnecessary and it's better to force the call of `peekMarkedByteSpan` as it is passed to the continuation. I'll change that.

**Comment by coot** [comment]
Are there any plans to release new version of `cborg` with this change included?
